hollow/src/main/java/com/netflix/hollow/api/producer/validation/ObjectModificationValidator.java

```diff
@@ -0,0 +1,122 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.hollow.api.producer.validation;
+
+import com.netflix.hollow.api.objects.HollowObject;
+import com.netflix.hollow.api.objects.generic.GenericHollowObject;
+import com.netflix.hollow.api.producer.HollowProducer.ReadState;
+import com.netflix.hollow.api.producer.HollowProducer.Validator;
+import com.netflix.hollow.api.producer.HollowProducerListener.Status;
+import com.netflix.hollow.core.HollowConstants;
+import com.netflix.hollow.core.index.HollowPrimaryKeyIndex;
+import com.netflix.hollow.core.index.key.PrimaryKey;
+import com.netflix.hollow.core.read.engine.HollowTypeReadState;
+import com.netflix.hollow.core.read.engine.object.HollowObjectTypeReadState;
+import com.netflix.hollow.core.schema.HollowSchema.SchemaType;
+import java.util.Arrays;
+import java.util.BitSet;
+import java.util.function.BiFunction;
+
+/**
+ * This validator runs a provided {@link BiFunction} on every instance of an object that was
+ * modified in a cycle. This "filter" function is called with the old GenericHollowObject as the
+ * first argument and the new GenericHollowObject as the second argument. If the function returns
+ * false, this validator is marked as failed and the producer cycle fails.
+ * Note that the filter function is not called with any objects that were added or removed.
+ * Also note that this validator can only be used on types that have a primary key.
+ */
+public class ObjectModificationValidator<T extends HollowObject> implements Validator {
+    private final String typeName;
+    private final BiFunction<T, T, Boolean> filter;
+
+    private SingleValidationStatus.SingleValidationStatusBuilder statusBuilder;
+
+    @SuppressWarnings("WeakerAccess")
+    public ObjectModificationValidator(String typeName, BiFunction<T, T, Boolean> filter) {
```

Comment:
`BiPredicate<? super T, ? super T> filter`, the function is consuming instances of T so can accept T or a super type.